build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/transport/GenerateTransportVersionDefinitionTask.java

```diff
@@ -0,0 +1,236 @@
+import org.gradle.api.provider.Property;
+import org.gradle.api.services.ServiceReference;
+import org.gradle.api.tasks.Input;
+import org.gradle.api.tasks.InputFiles;
+import org.gradle.api.tasks.Optional;
+import org.gradle.api.tasks.PathSensitive;
+import org.gradle.api.tasks.PathSensitivity;
+import org.gradle.api.tasks.TaskAction;
+import org.gradle.api.tasks.options.Option;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * This task generates transport version definition files. These files
+ * are runtime resources that TransportVersion loads statically.
+ * They contain a comma separated list of integer ids. Each file is named the same
+ * as the transport version name itself (with the .csv suffix).
+ */
+public abstract class GenerateTransportVersionDefinitionTask extends DefaultTask {
+
+    /**
+     * Files that contain the references to the transport version names.
+     */
+    @InputFiles
+    @PathSensitive(PathSensitivity.RELATIVE)
+    public abstract ConfigurableFileCollection getReferencesFiles();
+
+    @ServiceReference("transportVersionResources")
+    abstract Property<TransportVersionResourcesService> getResourceService();
+
+    @Input
+    @Optional
+    @Option(option = "name", description = "The name of the Transport Version definition, e.g. --name=my_new_tv")
+    public abstract Property<String> getTransportVersionName();
+
+    @Input
+    @Optional
+    @Option(
+        option = "backport-branches",
+        description = "The branches this definition will be backported to, e.g. --backport-branches=9.1,8.19"
+    )
+    public abstract Property<String> getBackportBranches();
+
+    @Input
+    @Optional
+    @Option(option = "increment", description = "The amount to increment the primary id for the main releaseBranch")
+    public abstract Property<Integer> getPrimaryIncrement();
+
+    /**
+     * The version number currently associated with the `main` branch, (e.g. as returned by
+     * VersionProperties.getElasticsearchVersion()).
+     */
+    @Input
+    public abstract Property<String> getMainReleaseBranch();
+
+    @TaskAction
+    public void run() throws IOException {
+        TransportVersionResourcesService resources = getResourceService().get();
+        Set<String> referencedNames = TransportVersionReference.collectNames(getReferencesFiles());
+        List<String> changedDefinitionNames = resources.getChangedReferableDefinitionNames();
+        String name = getTargetName(resources, referencedNames, changedDefinitionNames);
+
+        List<TransportVersionUpperBound> mainUpperBounds = resources.getUpperBoundsFromMain();
+        Set<String> releaseBranches = getTargetReleaseBranches();
+        checkReleaseBranches(mainUpperBounds, releaseBranches);
+
+        if (name.isEmpty()) {
+            resetAllUpperBounds(resources);
```

Comment:
I find that having this task both generate definitions _and_ update upper bounds to be a bit confusing and muddling of concerns. But I get that they both sort of use the same inputs so it makes sense to combine them.

Reply:
It really needs to do both because at the end of the task all the resources need to be consistent.